{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d5b072d4-58f0-4717-9a65-09234414722a",
   "metadata": {},
   "source": [
    "# Hands on introduction to ancient microbiome analysis\n",
    "Author: Maxime Borry  \n",
    "Date: 12/08/2021"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfb0e993-779a-45a1-9789-142d19d99243",
   "metadata": {},
   "source": [
    "In this tutorial, we're going to go through the steps necessary to:\n",
    "- generate a taxonomic profile table with [metaphlan v3.0](https://github.com/biobakery/MetaPhlAn)\n",
    "- have a look at metaphlan results with [Pavian](https://github.com/fbreitwieser/pavian) and generate a [Krona plot](https://github.com/marbl/Krona/wiki)\n",
    "- Compare the alpha diversity of our samples vs the diversity of modern human gut samples\n",
    "- Compare the composition of our samples vs modern gut samples, and see where they fit in a lower dimensional space"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1303742-de95-45ce-a244-ff6a1b76197a",
   "metadata": {},
   "source": [
    "## 0. Quick intro to Jupyter Notebooks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0497247e-225c-4b85-8fd1-6bc18480334b",
   "metadata": {},
   "source": [
    "This a markdown cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c9ab4799-42ad-4a55-8f27-86a515fbdc64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is a python cell\n"
     ]
    }
   ],
   "source": [
    "print(\"This is a python cell\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c68889c4-79fd-45ca-a0a1-1366f007b6b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is how to run a single line bash command\n"
     ]
    }
   ],
   "source": [
    "! echo \"This is how to run a single line bash command\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b0020257-63ae-40fc-8036-b6698a814a7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This how to run a multi\n",
      "lines bash command\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "echo \"This how to run a multi\"\n",
    "echo \"lines bash command\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaf4349f-bb50-4a17-b445-fc31fac3ecef",
   "metadata": {},
   "source": [
    "## 1. Data pre-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff616757-f604-468e-8f56-0e2563b51427",
   "metadata": {},
   "source": [
    "Before starting to analyze our data, we will need to pre-process them to remove reads mapping to the host genome, here, *Homo sapiens*\n",
    "\n",
    "To do so, I've used [nf-core/eager](https://github.com/nf-core/eager)\n",
    "\n",
    "I've already pre-processed the data, and the resulting cleaned files are available in the [data/eager_cleaned](data/eager_cleaned), but the basic eager command to do so is:\n",
    "\n",
    "` nextflow run nf-core/eager -profile <docker/singularity/podman/conda/institute> --input '*_R{1,2}.fastq.gz' --fasta 'human_genome.fasta' --hostremoval_input_fastq`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4faf5c98-266e-4718-9d61-a5016a0e60af",
   "metadata": {},
   "source": [
    "## 2. Adapter sequence trimming and low-quality bases trimming\n",
    "\n",
    "Sequencing adapters are small DNA sequences adding prior to DNA sequencing to allow the DNA fragments to attach to the sequencing flow cells. Because these adapters could interfere with downtstream analyses, we need to remove them before proceeding any further.\n",
    "Furthermore, because the quality of the sequencing is not always optimal, we need to remove bases of lower sequencing quality to might lead to spurious results in downstream analyses.\n",
    "\n",
    "To perform both of these tasks, we'll use the [AdapterRemoval](https://github.com/MikkelSchubert/adapterremoval) program."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "5a6f9943-d428-4821-942c-e0d1b6f30fe5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AdapterRemoval ver. 2.3.2\n",
      "\n",
      "This program searches for and removes remnant adapter sequences from\n",
      "your read data.  The program can analyze both single end and paired end\n",
      "data.  For detailed explanation of the parameters, please refer to the\n",
      "man page.  For comments, suggestions  and feedback please contact Stinus\n",
      "Lindgreen (stinus@binf.ku.dk) and Mikkel Schubert (MikkelSch@gmail.com).\n",
      "\n",
      "If you use the program, please cite the paper:\n",
      "    Schubert, Lindgreen, and Orlando (2016). AdapterRemoval v2: rapid\n",
      "    adapter trimming, identification, and read merging.\n",
      "    BMC Research Notes, 12;9(1):88.\n",
      "\n",
      "    http://bmcresnotes.biomedcentral.com/articles/10.1186/s13104-016-1900-2\n",
      "\n",
      "\n",
      "OPTIONS:\n",
      "    --help  \n",
      "        Display this message.\n",
      "\n",
      "    --version  \n",
      "        Print the version string.\n",
      "\n",
      "\n",
      "    --file1 FILE [FILE ...]\n",
      "        Input files containing mate 1 reads or single-ended reads; one or\n",
      "        more files may be listed [REQUIRED].\n",
      "\n",
      "    --file2 [FILE ...]\n",
      "        Input files containing mate 2 reads; if used, then the same number of\n",
      "        files as --file1 must be listed [OPTIONAL].\n",
      "\n",
      "    --identify-adapters  \n",
      "        Attempt to identify the adapter pair of PE reads, by searching for\n",
      "        overlapping mate reads [default: off].\n",
      "\n",
      "    --threads THREADS\n",
      "        Maximum number of threads [default: 1]\n",
      "\n",
      "\n",
      "FASTQ OPTIONS:\n",
      "    --qualitybase BASE\n",
      "        Quality base used to encode Phred scores in input; either 33, 64, or\n",
      "        solexa [default: 33].\n",
      "\n",
      "    --qualitybase-output BASE\n",
      "        Quality base used to encode Phred scores in output; either 33, 64, or\n",
      "        solexa. By default, reads will be written in the same format as the\n",
      "        that specified using --qualitybase.\n",
      "\n",
      "    --qualitymax BASE\n",
      "        Specifies the maximum Phred score expected in input files, and used\n",
      "        when writing output. ASCII encoded values are limited to the\n",
      "        characters '!' (ASCII = 33) to '~' (ASCII = 126), meaning that\n",
      "        possible scores are 0 - 93 with offset 33, and 0 - 62 for offset 64\n",
      "        and Solexa scores [default: 41].\n",
      "\n",
      "    --mate-separator CHAR\n",
      "        Character separating the mate number (1 or 2) from the read name in\n",
      "        FASTQ records [default: '/'].\n",
      "\n",
      "    --interleaved  \n",
      "        This option enables both the --interleaved-input option and the\n",
      "        --interleaved-output option [default: off].\n",
      "\n",
      "    --interleaved-input  \n",
      "        The (single) input file provided contains both the mate 1 and mate 2\n",
      "        reads, one pair after the other, with one mate 1 reads followed by\n",
      "        one mate 2 read. This option is implied by the --interleaved option\n",
      "        [default: off].\n",
      "\n",
      "    --interleaved-output  \n",
      "        If set, trimmed paired-end reads are written to a single file\n",
      "        containing mate 1 and mate 2 reads, one pair after the other. This\n",
      "        option is implied by the --interleaved option [default: off].\n",
      "\n",
      "    --combined-output  \n",
      "        If set, all reads are written to the same file(s), specified by\n",
      "        --output1 and --output2 (--output1 only if --interleaved-output is\n",
      "        not set). Discarded reads are replaced with a single 'N' with Phred\n",
      "        score 0 [default: off].\n",
      "\n",
      "\n",
      "OUTPUT FILES:\n",
      "    --basename BASENAME\n",
      "        Default prefix for all output files for which no filename was\n",
      "        explicitly set [default: your_output].\n",
      "\n",
      "    --settings FILE\n",
      "        Output file containing information on the parameters used in the run\n",
      "        as well as overall statistics on the reads after trimming [default:\n",
      "        BASENAME.settings]\n",
      "\n",
      "    --output1 FILE\n",
      "        Output file containing trimmed mate1 reads [default:\n",
      "        BASENAME.pair1.truncated (PE), BASENAME.truncated (SE), or\n",
      "        BASENAME.paired.truncated (interleaved PE)]\n",
      "\n",
      "    --output2 FILE\n",
      "        Output file containing trimmed mate 2 reads [default:\n",
      "        BASENAME.pair2.truncated (only used in PE mode, but not if\n",
      "        --interleaved-output is enabled)]\n",
      "\n",
      "    --singleton FILE\n",
      "        Output file to which containing paired reads for which the mate has\n",
      "        been discarded [default: BASENAME.singleton.truncated]\n",
      "\n",
      "    --outputcollapsed FILE\n",
      "        If --collapsed is set, contains overlapping mate-pairs which have\n",
      "        been merged into a single read (PE mode) or reads for which the\n",
      "        adapter was identified by a minimum overlap, indicating that the\n",
      "        entire template molecule is present. This does not include which have\n",
      "        subsequently been trimmed due to low-quality or ambiguous nucleotides\n",
      "        [default: BASENAME.collapsed]\n",
      "\n",
      "    --outputcollapsedtruncated FILE\n",
      "        Collapsed reads (see --outputcollapsed) which were trimmed due the\n",
      "        presence of low-quality or ambiguous nucleotides [default:\n",
      "        BASENAME.collapsed.truncated]\n",
      "\n",
      "    --discarded FILE\n",
      "        Contains reads discarded due to the --minlength, --maxlength or\n",
      "        --maxns options [default: BASENAME.discarded]\n",
      "\n",
      "\n",
      "OUTPUT COMPRESSION:\n",
      "    --gzip  \n",
      "        Enable gzip compression [default: off]\n",
      "\n",
      "    --gzip-level LEVEL\n",
      "        Compression level, 0 - 9 [default: 6]\n",
      "\n",
      "    --bzip2  \n",
      "        Enable bzip2 compression [default: off]\n",
      "\n",
      "    --bzip2-level LEVEL\n",
      "        Compression level, 0 - 9 [default: 9]\n",
      "\n",
      "\n",
      "TRIMMING SETTINGS:\n",
      "    --adapter1 SEQUENCE\n",
      "        Adapter sequence expected to be found in mate 1 reads [default:\n",
      "        AGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNATCTCGTATGCCGTCTTCTGCTTG].\n",
      "\n",
      "    --adapter2 SEQUENCE\n",
      "        Adapter sequence expected to be found in mate 2 reads [default:\n",
      "        AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT].\n",
      "\n",
      "    --adapter-list FILENAME\n",
      "        Read table of white-space separated adapters pairs, used as if the\n",
      "        first column was supplied to --adapter1, and the second column was\n",
      "        supplied to --adapter2; only the first adapter in each pair is\n",
      "        required SE trimming mode [default: <not set>].\n",
      "\n",
      "\n",
      "    --minadapteroverlap LENGTH\n",
      "        In single-end mode, reads are only trimmed if the overlap between\n",
      "        read and the adapter is at least X bases long, not counting ambiguous\n",
      "        nucleotides (N); this is independent of the --minalignmentlength when\n",
      "        using --collapse, allowing a conservative selection of putative\n",
      "        complete inserts while ensuring that all possible adapter\n",
      "        contamination is trimmed [default: 0].\n",
      "\n",
      "    --mm MISMATCH_RATE\n",
      "        Max error-rate when aligning reads and/or adapters. If > 1, the max\n",
      "        error-rate is set to 1 / MISMATCH_RATE; if < 0, the defaults are\n",
      "        used, otherwise the user-supplied value is used directly [default:\n",
      "        1/3 for trimming; 1/10 when identifying adapters].\n",
      "\n",
      "    --shift N\n",
      "        Consider alignments where up to N nucleotides are missing from the 5'\n",
      "        termini [default: 2].\n",
      "\n",
      "\n",
      "    --trim5p N [N]\n",
      "        Trim the 5' of reads by a fixed amount after removing adapters, but\n",
      "        before carrying out quality based trimming. Specify one value to trim\n",
      "        mate 1 and mate 2 reads the same amount, or two values separated by a\n",
      "        space to trim each mate different amounts [default: no trimming].\n",
      "\n",
      "    --trim3p N [N]\n",
      "        Trim the 3' of reads by a fixed amount. See --trim5p.\n",
      "\n",
      "    --trimns  \n",
      "        If set, trim ambiguous bases (N) at 5'/3' termini [default: off]\n",
      "\n",
      "    --maxns MAX\n",
      "        Reads containing more ambiguous bases (N) than this number after\n",
      "        trimming are discarded [default: 1000].\n",
      "\n",
      "    --trimqualities  \n",
      "        If set, trim bases at 5'/3' termini with quality scores <= to\n",
      "        --minquality value [default: off]\n",
      "\n",
      "    --trimwindows INT\n",
      "        If set, quality trimming will be carried out using window based\n",
      "        approach, where windows with an average quality less than\n",
      "        --minquality will be trimmed. If >= 1, this value will be used as the\n",
      "        window size. If the value is < 1, the value will be multiplied with\n",
      "        the read length to determine a window size per read. If the resulting\n",
      "        window size is 0 or larger than the read length, the read length is\n",
      "        used as the window size. This option implies --trimqualities\n",
      "        [default: <not set>].\n",
      "\n",
      "    --minquality PHRED\n",
      "        Inclusive minimum; see --trimqualities for details [default: 2]\n",
      "\n",
      "    --preserve5p  \n",
      "        If set, bases at the 5p will not be trimmed by --trimns,\n",
      "        --trimqualities, and --trimwindows. Collapsed reads will not be\n",
      "        quality trimmed when this option is enabled [default: 5p bases are\n",
      "        trimmed]\n",
      "\n",
      "\n",
      "    --minlength LENGTH\n",
      "        Reads shorter than this length are discarded following trimming\n",
      "        [default: 15].\n",
      "\n",
      "    --maxlength LENGTH\n",
      "        Reads longer than this length are discarded following trimming\n",
      "        [default: 4294967295].\n",
      "\n",
      "\n",
      "READ MERGING:\n",
      "    --collapse  \n",
      "        When set, paired ended read alignments of --minalignmentlength or\n",
      "        more bases are combined into a single consensus sequence,\n",
      "        representing the complete insert, and written to either\n",
      "        basename.collapsed or basename.collapsed.truncated (if trimmed due to\n",
      "        low-quality bases following collapse); for single-ended reads,\n",
      "        putative complete inserts are identified as having at least\n",
      "        --minalignmentlength bases overlap with the adapter sequence, and are\n",
      "        written to the the same files [default: off].\n",
      "\n",
      "    --collapse-deterministic  \n",
      "        In standard --collapse mode, AdapterRemoval will randomly select one\n",
      "        of two different overlapping bases if these have the same quality\n",
      "        (otherwise it picks the highest quality base). With\n",
      "        --collapse-deterministic, AdapterRemoval will instead set such bases\n",
      "        to N. Setting this option also sets--collapse [default: off].\n",
      "\n",
      "    --collapse-conservatively  \n",
      "        Enables a more conservative merging algorithm inspired by fastq-join,\n",
      "        in which the higher quality score is picked for matching bases and\n",
      "        the max score minus the min score is picked for mismatching bases.\n",
      "        For more details, see the documentation. --seed and\n",
      "        --collapse-deterministic have no effect when this is enabled. Setting\n",
      "        this option also sets --collapse [default: off].\n",
      "\n",
      "    --minalignmentlength LENGTH\n",
      "        If --collapse is set, paired reads must overlap at least this number\n",
      "        of bases to be collapsed, and single-ended reads must overlap at\n",
      "        least this number of bases with the adapter to be considered complete\n",
      "        template molecules [default: 11].\n",
      "\n",
      "    --seed SEED\n",
      "        Sets the RNG seed used when choosing between bases with equal Phred\n",
      "        scores when --collapse is enabled. This option is not available if\n",
      "        more than one thread is used. If not specified, aseed is generated\n",
      "        using the current time.\n",
      "\n",
      "\n",
      "DEMULTIPLEXING:\n",
      "    --barcode-list FILENAME\n",
      "        List of barcodes or barcode pairs for single or double-indexed\n",
      "        demultiplexing. Note that both indexes should be specified for both\n",
      "        single-end and paired-end trimming, if double-indexed multiplexing\n",
      "        was used, in order to ensure that the demultiplexed reads can be\n",
      "        trimmed correctly [default: <not set>].\n",
      "\n",
      "    --barcode-mm N\n",
      "        Maximum number of mismatches allowed when counting mismatches in both\n",
      "        the mate 1 and the mate 2 barcode for paired reads.\n",
      "\n",
      "    --barcode-mm-r1 N\n",
      "        Maximum number of mismatches allowed for the mate 1 barcode; if not\n",
      "        set, this value is equal to the '--barcode-mm' value; cannot be\n",
      "        higher than the '--barcode-mm value'.\n",
      "\n",
      "    --barcode-mm-r2 N\n",
      "        Maximum number of mismatches allowed for the mate 2 barcode; if not\n",
      "        set, this value is equal to the '--barcode-mm' value; cannot be\n",
      "        higher than the '--barcode-mm value'.\n",
      "\n",
      "    --demultiplex-only  \n",
      "        Only carry out demultiplexing using the list of barcodes supplied\n",
      "        with --barcode-list. No other processing is done.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! AdapterRemoval -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7784dab4-5453-4c4d-9534-c3b87e8415e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trimming paired end reads ...\n",
      "Opening FASTQ file '../data/eager_cleaned/ZSM028_R1.fastq.gz', line numbers start at 1\n",
      "Opening FASTQ file '../data/eager_cleaned/ZSM028_R2.fastq.gz', line numbers start at 1\n",
      "Processed a total of 1,996,618 reads in 25.6s; 78,000 reads per second on average ...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "AdapterRemoval --file1 ../data/eager_cleaned/ZSM028_R1.fastq.gz \\\n",
    "    --file2 ../data/eager_cleaned/ZSM028_R2.fastq.gz \\\n",
    "    --basename ../results/adapterremoval/ZSM028 \\\n",
    "    --gzip \\\n",
    "    --minquality 20 \\\n",
    "    --trimqualities \\\n",
    "    --collapse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b874898-0590-4e6d-9265-4cd5ccb2e236",
   "metadata": {},
   "source": [
    "Since Metaphlan doesn't use the paired-end information, we'll combine all output files together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "12a2bddc-9a7a-4d86-b8db-aa0d9ed79cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat ../results/adapterremoval/*.collapsed.* ../results/adapterremoval/*.truncated.* ../results/adapterremoval/*.singleton.*  > ../results/adapterremoval/ZSM028.combined.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f881863-e66e-4201-9600-1fc9e6edccc8",
   "metadata": {},
   "source": [
    "## 3. Taxonomic profiling with Metaphlan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "be2d77c0-0865-48ff-b727-76d2e106b77b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: metaphlan --input_type {fastq,fasta,bowtie2out,sam} [--force]\n",
      "                 [--bowtie2db METAPHLAN_BOWTIE2_DB] [-x INDEX]\n",
      "                 [--bt2_ps BowTie2 presets] [--bowtie2_exe BOWTIE2_EXE]\n",
      "                 [--bowtie2_build BOWTIE2_BUILD] [--bowtie2out FILE_NAME]\n",
      "                 [--min_mapq_val MIN_MAPQ_VAL] [--no_map] [--tmp_dir]\n",
      "                 [--tax_lev TAXONOMIC_LEVEL] [--min_cu_len]\n",
      "                 [--min_alignment_len] [--add_viruses] [--ignore_eukaryotes]\n",
      "                 [--ignore_bacteria] [--ignore_archaea] [--stat_q]\n",
      "                 [--perc_nonzero] [--ignore_markers IGNORE_MARKERS]\n",
      "                 [--avoid_disqm] [--stat] [-t ANALYSIS TYPE]\n",
      "                 [--nreads NUMBER_OF_READS] [--pres_th PRESENCE_THRESHOLD]\n",
      "                 [--clade] [--min_ab] [-o output file] [--sample_id_key name]\n",
      "                 [--use_group_representative] [--sample_id value]\n",
      "                 [-s sam_output_file] [--legacy-output] [--CAMI_format_output]\n",
      "                 [--unknown_estimation] [--biom biom_output] [--mdelim mdelim]\n",
      "                 [--nproc N] [--install] [--force_download]\n",
      "                 [--read_min_len READ_MIN_LEN] [-v] [-h]\n",
      "                 [INPUT_FILE] [OUTPUT_FILE]\n",
      "\n",
      "DESCRIPTION\n",
      " MetaPhlAn version 3.0.13 (27 Jul 2021): \n",
      " METAgenomic PHyLogenetic ANalysis for metagenomic taxonomic profiling.\n",
      "\n",
      "AUTHORS: Francesco Beghini (francesco.beghini@unitn.it),Nicola Segata (nicola.segata@unitn.it), Duy Tin Truong, Francesco Asnicar (f.asnicar@unitn.it), Aitor Blanco Miguez (aitor.blancomiguez@unitn.it)\n",
      "\n",
      "COMMON COMMANDS\n",
      "\n",
      " We assume here that MetaPhlAn is installed using the several options available (pip, conda, PyPi)\n",
      " Also BowTie2 should be in the system path with execution and read permissions, and Perl should be installed)\n",
      "\n",
      "========== MetaPhlAn clade-abundance estimation ================= \n",
      "\n",
      "The basic usage of MetaPhlAn consists in the identification of the clades (from phyla to species ) \n",
      "present in the metagenome obtained from a microbiome sample and their \n",
      "relative abundance. This correspond to the default analysis type (-t rel_ab).\n",
      "\n",
      "*  Profiling a metagenome from raw reads:\n",
      "$ metaphlan metagenome.fastq --input_type fastq -o profiled_metagenome.txt\n",
      "\n",
      "*  You can take advantage of multiple CPUs and save the intermediate BowTie2 output for re-running\n",
      "   MetaPhlAn extremely quickly:\n",
      "$ metaphlan metagenome.fastq --bowtie2out metagenome.bowtie2.bz2 --nproc 5 --input_type fastq -o profiled_metagenome.txt\n",
      "\n",
      "*  If you already mapped your metagenome against the marker DB (using a previous MetaPhlAn run), you\n",
      "   can obtain the results in few seconds by using the previously saved --bowtie2out file and \n",
      "   specifying the input (--input_type bowtie2out):\n",
      "$ metaphlan metagenome.bowtie2.bz2 --nproc 5 --input_type bowtie2out -o profiled_metagenome.txt\n",
      "\n",
      "*  bowtie2out files generated with MetaPhlAn versions below 3 are not compatibile.\n",
      "   Starting from MetaPhlAn 3.0, the BowTie2 ouput now includes the size of the profiled metagenome and the average read length.\n",
      "   If you want to re-run MetaPhlAn using these file you should provide the metagenome size via --nreads:\n",
      "$ metaphlan metagenome.bowtie2.bz2 --nproc 5 --input_type bowtie2out --nreads 520000 -o profiled_metagenome.txt\n",
      "\n",
      "*  You can also provide an externally BowTie2-mapped SAM if you specify this format with \n",
      "   --input_type. Two steps: first apply BowTie2 and then feed MetaPhlAn with the obtained sam:\n",
      "$ bowtie2 --sam-no-hd --sam-no-sq --no-unal --very-sensitive -S metagenome.sam -x ${mpa_dir}/metaphlan_databases/mpa_v30_CHOCOPhlAn_201901 -U metagenome.fastq\n",
      "$ metaphlan metagenome.sam --input_type sam -o profiled_metagenome.txt\n",
      "\n",
      "*  We can also natively handle paired-end metagenomes, and, more generally, metagenomes stored in \n",
      "  multiple files (but you need to specify the --bowtie2out parameter):\n",
      "$ metaphlan metagenome_1.fastq,metagenome_2.fastq --bowtie2out metagenome.bowtie2.bz2 --nproc 5 --input_type fastq\n",
      "\n",
      "------------------------------------------------------------------- \n",
      " \n",
      "\n",
      "========== Marker level analysis ============================ \n",
      "\n",
      "MetaPhlAn introduces the capability of characterizing organisms at the strain level using non\n",
      "aggregated marker information. Such capability comes with several slightly different flavours and \n",
      "are a way to perform strain tracking and comparison across multiple samples.\n",
      "Usually, MetaPhlAn is first ran with the default -t to profile the species present in\n",
      "the community, and then a strain-level profiling can be performed to zoom-in into specific species\n",
      "of interest. This operation can be performed quickly as it exploits the --bowtie2out intermediate \n",
      "file saved during the execution of the default analysis type.\n",
      "\n",
      "*  The following command will output the abundance of each marker with a RPK (reads per kilo-base) \n",
      "   higher 0.0. (we are assuming that metagenome_outfmt.bz2 has been generated before as \n",
      "   shown above).\n",
      "$ metaphlan -t marker_ab_table metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt\n",
      "   The obtained RPK can be optionally normalized by the total number of reads in the metagenome \n",
      "   to guarantee fair comparisons of abundances across samples. The number of reads in the metagenome\n",
      "   needs to be passed with the '--nreads' argument\n",
      "\n",
      "*  The list of markers present in the sample can be obtained with '-t marker_pres_table'\n",
      "$ metaphlan -t marker_pres_table metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt\n",
      "   The --pres_th argument (default 1.0) set the minimum RPK value to consider a marker present\n",
      "\n",
      "*  The list '-t clade_profiles' analysis type reports the same information of '-t marker_ab_table'\n",
      "   but the markers are reported on a clade-by-clade basis.\n",
      "$ metaphlan -t clade_profiles metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt\n",
      "\n",
      "*  Finally, to obtain all markers present for a specific clade and all its subclades, the \n",
      "   '-t clade_specific_strain_tracker' should be used. For example, the following command\n",
      "   is reporting the presence/absence of the markers for the B. fragilis species and its strains\n",
      "   the optional argument --min_ab specifies the minimum clade abundance for reporting the markers\n",
      "\n",
      "$ metaphlan -t clade_specific_strain_tracker --clade s__Bacteroides_fragilis metagenome_outfmt.bz2 --input_type bowtie2out -o marker_abundance_table.txt\n",
      "\n",
      "------------------------------------------------------------------- \n",
      "\n",
      "positional arguments:\n",
      "  INPUT_FILE            the input file can be:\n",
      "                        * a fastq file containing metagenomic reads\n",
      "                        OR\n",
      "                        * a BowTie2 produced SAM file. \n",
      "                        OR\n",
      "                        * an intermediary mapping file of the metagenome generated by a previous MetaPhlAn run \n",
      "                        If the input file is missing, the script assumes that the input is provided using the standard \n",
      "                        input, or named pipes.\n",
      "                        IMPORTANT: the type of input needs to be specified with --input_type\n",
      "  OUTPUT_FILE           the tab-separated output file of the predicted taxon relative abundances \n",
      "                        [stdout if not present]\n",
      "\n",
      "Required arguments:\n",
      "  --input_type {fastq,fasta,bowtie2out,sam}\n",
      "                        set whether the input is the FASTA file of metagenomic reads or \n",
      "                        the SAM file of the mapping of the reads against the MetaPhlAn db.\n",
      "\n",
      "Mapping arguments:\n",
      "  --force               Force profiling of the input file by removing the bowtie2out file\n",
      "  --bowtie2db METAPHLAN_BOWTIE2_DB\n",
      "                        Folder containing the MetaPhlAn database. You can specify the location by exporting the DEFAULT_DB_FOLDER variable in the shell.[default /home/maxime_borry/.conda/envs/summer_school_microbiome/lib/python3.9/site-packages/metaphlan/metaphlan_databases]\n",
      "  -x INDEX, --index INDEX\n",
      "                        Specify the id of the database version to use. If \"latest\", MetaPhlAn will get the latest version.\n",
      "                        If an index name is provided, MetaPhlAn will try to use it, if available, and skip the online check.\n",
      "                        If the database files are not found on the local MetaPhlAn installation they\n",
      "                        will be automatically downloaded [default latest]\n",
      "  --bt2_ps BowTie2 presets\n",
      "                        Presets options for BowTie2 (applied only when a FASTA file is provided)\n",
      "                        The choices enabled in MetaPhlAn are:\n",
      "                         * sensitive\n",
      "                         * very-sensitive\n",
      "                         * sensitive-local\n",
      "                         * very-sensitive-local\n",
      "                        [default very-sensitive]\n",
      "  --bowtie2_exe BOWTIE2_EXE\n",
      "                        Full path and name of the BowTie2 executable. This option allowsMetaPhlAn to reach the executable even when it is not in the system PATH or the system PATH is unreachable\n",
      "  --bowtie2_build BOWTIE2_BUILD\n",
      "                        Full path to the bowtie2-build command to use, deafult assumes that 'bowtie2-build is present in the system path\n",
      "  --bowtie2out FILE_NAME\n",
      "                        The file for saving the output of BowTie2\n",
      "  --min_mapq_val MIN_MAPQ_VAL\n",
      "                        Minimum mapping quality value (MAPQ) [default 5]\n",
      "  --no_map              Avoid storing the --bowtie2out map file\n",
      "  --tmp_dir             The folder used to store temporary files [default is the OS dependent tmp dir]\n",
      "\n",
      "Post-mapping arguments:\n",
      "  --tax_lev TAXONOMIC_LEVEL\n",
      "                        The taxonomic level for the relative abundance output:\n",
      "                        'a' : all taxonomic levels\n",
      "                        'k' : kingdoms\n",
      "                        'p' : phyla only\n",
      "                        'c' : classes only\n",
      "                        'o' : orders only\n",
      "                        'f' : families only\n",
      "                        'g' : genera only\n",
      "                        's' : species only\n",
      "                        [default 'a']\n",
      "  --min_cu_len          minimum total nucleotide length for the markers in a clade for\n",
      "                        estimating the abundance without considering sub-clade abundances\n",
      "                        [default 2000]\n",
      "  --min_alignment_len   The sam records for aligned reads with the longest subalignment\n",
      "                        length smaller than this threshold will be discarded.\n",
      "                        [default None]\n",
      "  --add_viruses         Allow the profiling of viral organisms\n",
      "  --ignore_eukaryotes   Do not profile eukaryotic organisms\n",
      "  --ignore_bacteria     Do not profile bacterial organisms\n",
      "  --ignore_archaea      Do not profile archeal organisms\n",
      "  --stat_q              Quantile value for the robust average\n",
      "                        [default 0.2]\n",
      "  --perc_nonzero        Percentage of markers with a non zero relative abundance for misidentify a species\n",
      "                        [default 0.33]\n",
      "  --ignore_markers IGNORE_MARKERS\n",
      "                        File containing a list of markers to ignore. \n",
      "  --avoid_disqm         Deactivate the procedure of disambiguating the quasi-markers based on the \n",
      "                        marker abundance pattern found in the sample. It is generally recommended \n",
      "                        to keep the disambiguation procedure in order to minimize false positives\n",
      "  --stat                Statistical approach for converting marker abundances into clade abundances\n",
      "                        'avg_g'  : clade global (i.e. normalizing all markers together) average\n",
      "                        'avg_l'  : average of length-normalized marker counts\n",
      "                        'tavg_g' : truncated clade global average at --stat_q quantile\n",
      "                        'tavg_l' : truncated average of length-normalized marker counts (at --stat_q)\n",
      "                        'wavg_g' : winsorized clade global average (at --stat_q)\n",
      "                        'wavg_l' : winsorized average of length-normalized marker counts (at --stat_q)\n",
      "                        'med'    : median of length-normalized marker counts\n",
      "                        [default tavg_g]\n",
      "\n",
      "Additional analysis types and arguments:\n",
      "  -t ANALYSIS TYPE      Type of analysis to perform: \n",
      "                         * rel_ab: profiling a metagenomes in terms of relative abundances\n",
      "                         * rel_ab_w_read_stats: profiling a metagenomes in terms of relative abundances and estimate the number of reads coming from each clade.\n",
      "                         * reads_map: mapping from reads to clades (only reads hitting a marker)\n",
      "                         * clade_profiles: normalized marker counts for clades with at least a non-null marker\n",
      "                         * marker_ab_table: normalized marker counts (only when > 0.0 and normalized by metagenome size if --nreads is specified)\n",
      "                         * marker_counts: non-normalized marker counts [use with extreme caution]\n",
      "                         * marker_pres_table: list of markers present in the sample (threshold at 1.0 if not differently specified with --pres_th\n",
      "                         * clade_specific_strain_tracker: list of markers present for a specific clade, specified with --clade, and all its subclades\n",
      "                        [default 'rel_ab']\n",
      "  --nreads NUMBER_OF_READS\n",
      "                        The total number of reads in the original metagenome. It is used only when \n",
      "                        -t marker_table is specified for normalizing the length-normalized counts \n",
      "                        with the metagenome size as well. No normalization applied if --nreads is not \n",
      "                        specified\n",
      "  --pres_th PRESENCE_THRESHOLD\n",
      "                        Threshold for calling a marker present by the -t marker_pres_table option\n",
      "  --clade               The clade for clade_specific_strain_tracker analysis\n",
      "  --min_ab              The minimum percentage abundance for the clade in the clade_specific_strain_tracker analysis\n",
      "\n",
      "Output arguments:\n",
      "  -o output file, --output_file output file\n",
      "                        The output file (if not specified as positional argument)\n",
      "  --sample_id_key name  Specify the sample ID key for this analysis. Defaults to 'SampleID'.\n",
      "  --use_group_representative\n",
      "                        Use a species as representative for species groups.\n",
      "  --sample_id value     Specify the sample ID for this analysis. Defaults to 'Metaphlan_Analysis'.\n",
      "  -s sam_output_file, --samout sam_output_file\n",
      "                        The sam output file\n",
      "  --legacy-output       Old MetaPhlAn2 two columns output\n",
      "  --CAMI_format_output  Report the profiling using the CAMI output format\n",
      "  --unknown_estimation  Scale relative abundances to the number of reads mapping to known clades in order to estimate unknowness\n",
      "  --biom biom_output, --biom_output_file biom_output\n",
      "                        If requesting biom file output: The name of the output file in biom format \n",
      "  --mdelim mdelim, --metadata_delimiter_char mdelim\n",
      "                        Delimiter for bug metadata: - defaults to pipe. e.g. the pipe in k__Bacteria|p__Proteobacteria \n",
      "\n",
      "Other arguments:\n",
      "  --nproc N             The number of CPUs to use for parallelizing the mapping [default 4]\n",
      "  --install             Only checks if the MetaPhlAn DB is installed and installs it if not. All other parameters are ignored.\n",
      "  --force_download      Force the re-download of the latest MetaPhlAn database.\n",
      "  --read_min_len READ_MIN_LEN\n",
      "                        Specify the minimum length of the reads to be considered when parsing the input file with 'read_fastx.py' script, default value is 70\n",
      "  -v, --version         Prints the current MetaPhlAn version and exit\n",
      "  -h, --help            show this help message and exit\n"
     ]
    }
   ],
   "source": [
    "! metaphlan  --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "32e368df-af39-4b43-b8c9-e5e3edd5dbc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: The metagenome profile contains clades that represent multiple species merged into a single representant.\n",
      "An additional column listing the merged species is added to the MetaPhlAn output.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "metaphlan ../results/adapterremoval/ZSM028.combined.fastq.gz  \\\n",
    "    --input_type fastq \\\n",
    "    --bowtie2db /home/maxime_borry/.conda/envs/summer_school_microbiome/lib/python3.9/site-packages/metaphlan/metaphlan_databases/  \\\n",
    "    --bowtie2out ../results/metaphlan/ZSM028.bt2.out  \\\n",
    "    > ../results/metaphlan/ZSM028.metaphlan_profile.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb9e9dc4-9211-4417-9aed-4ed8eeec314d",
   "metadata": {},
   "source": [
    "The main results files that we're interested in is located at [../results/metaphlan/ZSM028.metaphlan_profile.txt](../results/metaphlan/ZSM028.metaphlan_profile.txt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6412c5b-166c-4302-b393-7a61a4986bf2",
   "metadata": {},
   "source": [
    "## 4. Visualizing metaphlan results with Pavian\n",
    "\n",
    "[Pavian](https://github.com/fbreitwieser/pavian) is an interactive app to explore results of different taxonomic classifiers\n",
    "\n",
    "There are different ways to run it:\n",
    "- If you have [docker](https://www.docker.com/) installed (and are somehow familiar with it)\n",
    "\n",
    "```bash\n",
    "docker pull 'florianbw/pavian'\n",
    "docker run --rm -p 5000:80 florianbw/pavian\n",
    "```\n",
    "\n",
    "Then open your browser and visit [localhost:5000](localhost:5000)\n",
    "\n",
    "- If you are familiar with [R](https://www.r-project.org/)\n",
    "\n",
    "```r\n",
    "if (!require(remotes)) { install.packages(\"remotes\") }\n",
    "remotes::install_github(\"fbreitwieser/pavian\")\n",
    "\n",
    "pavian::runApp(port=5000)\n",
    "```\n",
    "\n",
    "Then open your browser and visit [localhost:5000](localhost:5000)\n",
    "\n",
    "- Otherwise, just visit [fbreitwieser.shinyapps.io/pavian](https://fbreitwieser.shinyapps.io/pavian/.)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4040d591-385e-460f-9216-a5d9e055e5f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-summer_school_microbiome]",
   "language": "python",
   "name": "conda-env-.conda-summer_school_microbiome-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
